utopia-vscode-extension/src/in-mem-fs.ts

```diff
@@ -0,0 +1,400 @@
+  const maybeExistingFile = getItem(path)
+  if (maybeExistingFile != null && isDirectory(maybeExistingFile)) {
+    throw isDirectoryError(path)
+  }
+
+  const now = Date.now()
+  const fileCTime = maybeExistingFile == null ? now : maybeExistingFile.ctime
+  const lastSavedTime =
+    unsavedContent == null || maybeExistingFile == null ? now : maybeExistingFile.lastSavedTime
+  const fileToWrite = fsFile(content, unsavedContent, fileCTime, now, lastSavedTime)
+  setItem(path, fileToWrite)
+  if (parent != null) {
+    markModified(parent)
+  }
+}
+
+export function writeFileSavedContent(path: string, content: Uint8Array) {
+  writeFile(path, content, null)
+}
+
+export function writeFileUnsavedContent(path: string, unsavedContent: Uint8Array) {
+  const savedContent = readFileSavedContent(path)
+  writeFile(path, savedContent, unsavedContent)
+}
+
+export function writeFileAsUTF8(path: string, content: string, unsavedContent: string | null) {
+  writeFile(
+    path,
+    encoder.encode(content),
+    unsavedContent == null ? null : encoder.encode(unsavedContent),
+  )
+}
+
+export function writeFileSavedContentAsUTF8(path: string, savedContent: string) {
+  writeFileAsUTF8(path, savedContent, null)
+}
+
+export function writeFileUnsavedContentAsUTF8(path: string, unsavedContent: string) {
+  writeFileUnsavedContent(path, encoder.encode(unsavedContent))
+}
+
+export function clearFileUnsavedContent(path: string) {
+  const savedContent = readFileSavedContent(path)
+  writeFileSavedContent(path, savedContent)
+}
+
+function updateMTime(node: FSNode): FSNode {
+  const now = Date.now()
+  if (isFile(node)) {
+    const lastSavedTime = node.unsavedContent == null ? now : node.lastSavedTime
+    return fsFile(node.content, node.unsavedContent, node.ctime, now, lastSavedTime)
+  } else {
+    return fsDirectory(node.ctime, now)
+  }
+}
+
+function markModified(nodeWithPath: FSNodeWithPath) {
+  setItem(nodeWithPath.path, updateMTime(nodeWithPath.node))
+}
+
+function uncheckedMove(oldPath: string, newPath: string) {
+  const node = getNode(oldPath)
+  setItem(newPath, updateMTime(node))
+  removeItem(oldPath)
+}
+
+export function rename(oldPath: string, newPath: string) {
+  const oldParent = getParent(oldPath)
+  const newParent = getParent(newPath)
+
+  const pathsToMove = targetsForOperation(oldPath, true)
+  const toNewPath = (p: string) => `${newPath}${p.slice(0, oldPath.length)}`
+  pathsToMove.forEach((pathToMove) => uncheckedMove(pathToMove, toNewPath(pathToMove)))
+  if (oldParent != null) {
+    markModified(oldParent)
+  }
+  if (newParent != null) {
+    markModified(newParent)
+  }
+}
+
+export function deletePath(path: string, recursive: boolean) {
+  const parent = getParent(path)
+  const targets = targetsForOperation(path, recursive)
+
+  // Really this should fail if recursive isn't set to true when trying to delete a
+  // non-empty directory, but for some reason VSCode doesn't provide an error suitable for that
```

Comment:
maybe throwing an error is the expected behavior?

Reply:
It still needs to tell VS Code why it failed though, which is why this is awkward